Fix dynamic tabs registration and a11y in NcAppSidebarTabs with provide/inject

ShGKme opened the pull request:
Fixes: #3461

Currently, `NcAppSidebarTabs` gets content passed slot, goes throw all nodes, and detects `NcAppSidebarTab` instances. But it happened once on mounted and does not support tabs update.

In this solution:
1. `NcAppSidebarTabs` provides:
   - `registerTab` method
   - `unregisterTab` method
   - `activeTab` property (replacement for `$parent`), could be reverted
2. All `NcAppSidebarTab` register and unregister themself on `created` and `beforeDestroy`
3. After registration `NcAppSidebarTabs` has the tab's data directly from the `NcAppSidebarTab`s

Sorting works the same way it was before. But there is no warnings if a user passes not-tabs as content.

This solution is fully compatible with Vue 3 as it is and Composition API.

![1](https://user-images.githubusercontent.com/25978914/225331451-c4b0f99f-9307-4252-85b3-8199f290d0e0.gif)

![2](https://user-images.githubusercontent.com/25978914/225331481-df7036d3-570e-4a0d-9bed-77a4e6ccd849.gif)

raimund-schluessler commented on src/components/NcAppSidebar/NcAppSidebar.vue:
```diff
@@ -57,10 +66,87 @@ include a standard-header like it's used by the files app.
+</template>
+<script>
+import Cog from 'vue-material-design-icons/Cog'
+import ShareVariant from 'vue-material-design-icons/ShareVariant'
```
```suggestion
```
Unused.

raimund-schluessler commented on src/components/NcAppSidebar/NcAppSidebar.vue:
```diff
@@ -57,10 +66,87 @@ include a standard-header like it's used by the files app.
+export default {
+	components: {
+		Cog,
+		ShareVariant,
```
```suggestion
```

raimund-schluessler commented on src/components/NcAppSidebar/NcAppSidebar.vue:
````diff
@@ -57,10 +66,87 @@ include a standard-header like it's used by the files app.
+```vue
+<template>
+	<div>
+		<input type="checkbox" v-model="hideFirstTab"> The first tab is hidden
````
```suggestion
		<NcCheckboxRadioSwitch :checked.sync="hideFirstTab">The first tab is hidden</NcCheckboxRadioSwitch>
```

raimund-schluessler commented on src/components/NcAppSidebar/NcAppSidebar.vue:
```diff
@@ -31,22 +31,31 @@ include a standard-header like it's used by the files app.
-		</NcAppSidebarTab>
-	</NcAppSidebar>
+	<div>
+		<input type="checkbox" v-model="hideFirstTab"> The first tab is hidden
```
```suggestion
		<NcCheckboxRadioSwitch :checked.sync="hideFirstTab">The first tab is hidden</NcCheckboxRadioSwitch>
```

raimund-schluessler commented on src/components/NcAppSidebar/NcAppSidebarTabs.vue:
```diff
@@ -216,64 +210,38 @@ export default {
 		updateActive() {
 			this.activeTab = this.active
-				&& this.tabs.findIndex(tab => tab.id === this.active) !== -1
+			&& this.tabs.findIndex(tab => tab.id === this.active) !== -1
```
```suggestion
			&& this.tabs.some(tab => tab.id === this.active)
```

raimund-schluessler commented on src/components/NcAppSidebar/NcAppSidebarTabs.vue:
```diff
@@ -121,6 +117,16 @@ export default {
 		currentTabIndex() {
 			return this.tabs.findIndex(tab => tab.id === this.activeTab)
```
```suggestion
			return this.sortedTabs.findIndex(tab => tab.id === this.activeTab)
```
The keyboard navigation is currently messed up because of this.

raimund-schluessler replied:
It also needs to be adjusted in the `focus*Tab` methods below.

raimund-schluessler commented on src/components/NcAppSidebarTab/NcAppSidebarTab.vue:
```diff
@@ -68,12 +70,19 @@ export default {
-		// TODO: implement a better way to force pass a prop fromm Sidebar
 		isActive() {
-			return this.$parent.activeTab === this.id
+			return this.activeTab === this.id
```
:+1:

raimund-schluessler replied:
I just realized that this doesn't work in the Tasks app. I have no idea why, though.

raimund-schluessler replied:
@ShGKme I think it would be a good idea to revert this change for now. We can still use it when we figure out why it's broken currently. For vue3 it works fine, though.

ShGKme replied:
It could be a problem with getter is old Vue is used. Updated with classic getter instead of `computed`.

raimund-schluessler commented on src/components/NcAppSidebar/NcAppSidebarTabs.vue:
```diff
@@ -216,64 +210,38 @@ export default {
-				return orderA - orderB
-			})
+		registerTab(tab) {
+			this.tabs.push(tab)
```
We could still validate the tabs before pushing them to the `tabs` array like above:
```js
				if (IsValidString(tab?.name)
					&& IsValidStringWithoutSpaces(tab?.id)
					&& (IsValidStringWithoutSpaces(tab?.icon) || tab?.$slots?.icon))
```

raimund-schluessler replied:
We could do the sorting here already. Then we don't need the computed prop.

raimund-schluessler reviewed (changes requested):
I like the approach and I think that we can waive the warning when tabs and non-tabs are mixed. The keyboard navigation needs fixing, but that's an easy fix.

raimund-schluessler commented:
It currently does not work in the Tasks app, and I didn't figure out yet why.

raimund-schluessler commented:
> It currently does not work in the Tasks app, and I didn't figure out yet why.

For some reason, the injected `activeTab` value is not reactive when I test it in the Tasks app. Reverting to `return this.$parent.activeTab === this.id` fixes this. But I have no idea why though, especially since it works in the netlify docs. :shrug:

raimund-schluessler commented:
@ShGKme I pushed a few commits to fix some issues that I found, because I wanted to test the changes with my vue3 branch. If you don't like the changes, feel free to discard them.

raimund-schluessler commented:
I checked that the approach also works for vue3. A few adjustments are of course necessary, but it works fine for vue3 as well, see https://github.com/nextcloud/nextcloud-vue/pull/3692.

raimund-schluessler commented:
@ShGKme Ping :slightly_smiling_face:

ShGKme commented:
> @ShGKme Ping 🙂

Pong 👀. I'll work on this today.

ShGKme commented:
Rebased onto master and rewrote WIP commit

ShGKme commented:
## Finished fix

- Removed mixing tabs and not tabs content test
- Replace `computed` with classic getter in case that was the problem with the Tasks app
- Fix activating tabs after registration/unregistration (active prop in prior)
- `NcAppSidebarTab`'s icon slot rendering method moved to `NcAppSidebarTab` as a public Tab's method
- `NcAppSidebarTab`'s icon slot default content rendering also moved to `NcAppSidebarTab` component (otherwise slot and default slot content were defined in a different components).
- All exposed `NcAppSidebarTab` instance properties defined in `expose`

## Fix a11y

- Set active tab's `tabindex` to `0` instead of `undefined`. It allows to navigate to tabs with keyboard (`Tab`, `SHIFT+Tab`). Previously you could move from tab to tab's content but couldn't move back...
- Handle `Home` and `End` keypress to navigate to the first/last tab according to w3

See: https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html

@raimund-schluessler Don't you know if previous behavior was intended? It wasn't possible to focus tabs navigation with `Tab` of `Shift+Tab`.

![Tab](https://user-images.githubusercontent.com/25978914/228106881-b49e6dc2-a77a-4944-8bb4-2338c8263fb9.gif)

## Updated docs with JSDocs for events and default slots and examples

![DynamicTabs](https://user-images.githubusercontent.com/25978914/228105469-1dd9ffae-71d7-4d75-9eee-bc5c7b347008.gif)

![Activating](https://user-images.githubusercontent.com/25978914/228105476-fd1bfa30-fd2a-448c-8ad7-0d11ee1d6c1b.gif)

![CustomOrder](https://user-images.githubusercontent.com/25978914/228105486-73479071-6b7c-447e-91d9-b5fc1cd59664.png)

ShGKme commented:
Tested in Talk

![TabsTalk](https://user-images.githubusercontent.com/25978914/228109069-bddc35fd-e68d-4bf3-ac7b-f426b7877d6c.gif)

skjnldsv reviewed (approved):
Awesome work!

raimund-schluessler reviewed (approved):
Works nicely now with Tasks as well!

raimund-schluessler commented:
Shall we squash the commits or merge as is?

ShGKme commented:
> Shall we squash the commits or merge as is?

I think, we may squash commits between the first `WIP` and finished fixing issue in `fix dynamic tabs registration`. But the last two other commits are less related to the original problem of the PR: a11y and general docs, I'd keep them separated.

I squashed with `Co-authored-by`, hope GH understands this :D
